File: crates/murl-core/src/canonical.rs
//! mURL Canonical Form 1 (MCF-1) for JSON documents.
//!
//! Signatures and integrity hashes need a deterministic byte representation
//! of a manifest. MCF-1 is a deliberately small canonicalization:
//!
//! * Object members sorted by Unicode code point of the member name
//!   (byte order of UTF-8, which is identical).
//! * No insignificant whitespace.
//! * String escaping: `\"`, `\\`, `\b`, `\f`, `\n`, `\r`, `\t`, and `\u00XX`
//!   for remaining control characters; all other characters emitted as raw
//!   UTF-8.
//! * Numbers MUST be integers representable in `i64`/`u64`, emitted in
//!   decimal with no exponent, no fraction, no leading zeros. Non-integer
//!   numbers are a hard error.
//!
//! Relation to RFC 8785 (JCS): MCF-1 output is byte-identical to JCS for
//! every document the manifest schema allows (ASCII member names, integer
//! numbers). It differs from JCS in two deliberate restrictions: member names
//! are sorted by code point rather than UTF-16 code unit (identical for BMP
//! member names, which the schema requires in practice), and non-integer
//! numbers are rejected instead of being formatted with ECMAScript rules.
//! Rejecting floats removes the single hardest-to-reimplement part of JCS and
//! with it an entire class of cross-implementation signature mismatches.

use serde_json::Value;
use thiserror::Error;

#[derive(Debug, Error, PartialEq, Eq)]
pub enum CanonicalError {
    #[error("canonical form forbids non-integer numbers (found {0})")]
    NonIntegerNumber(String),
}

/// Serialize a JSON value into MCF-1 canonical bytes.
pub fn canonical_json_bytes(value: &Value) -> Result<Vec<u8>, CanonicalError> {
    let mut out = String::new();
    write_value(value, &mut out)?;
    Ok(out.into_bytes())
}

fn write_value(value: &Value, out: &mut String) -> Result<(), CanonicalError> {
    match value {
        Value::Null => out.push_str("null"),
        Value::Bool(true) => out.push_str("true"),
        Value::Bool(false) => out.push_str("false"),
        Value::Number(n) => {
            if let Some(i) = n.as_i64() {
                out.push_str(&i.to_string());
            } else if let Some(u) = n.as_u64() {
                out.push_str(&u.to_string());
            } else {
                return Err(CanonicalError::NonIntegerNumber(n.to_string()));
            }
        }
        Value::String(s) => write_string(s, out),
        Value::Array(items) => {
            out.push('[');
            for (i, item) in items.iter().enumerate() {
                if i > 0 {
                    out.push(',');
                }
                write_value(item, out)?;
            }
            out.push(']');
        }
        Value::Object(map) => {
            // serde_json's default Map is a BTreeMap: iteration is already
            // sorted by key (byte order == code point order for UTF-8).
            out.push('{');
            for (i, (k, v)) in map.iter().enumerate() {
                if i > 0 {
                    out.push(',');
                }
                write_string(k, out);
                out.push(':');
                write_value(v, out)?;
            }
            out.push('}');
        }
    }
    Ok(())
}

fn write_string(s: &str, out: &mut String) {
    out.push('"');
    for c in s.chars() {
        match c {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\u{0008}' => out.push_str("\\b"),
            '\u{000c}' => out.push_str("\\f"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            c if (c as u32) < 0x20 => {
                out.push_str(&format!("\\u{:04x}", c as u32));
            }
            c => out.push(c),
        }
    }
    out.push('"');
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn sorts_keys_and_strips_whitespace() {
        let v: Value =
            serde_json::from_str(r#"{ "b": 1, "a": { "z": [1, 2], "y": null } }"#).unwrap();
        let c = String::from_utf8(canonical_json_bytes(&v).unwrap()).unwrap();
        assert_eq!(c, r#"{"a":{"y":null,"z":[1,2]},"b":1}"#);
    }

    #[test]
    fn escapes_exactly_like_jcs() {
        let v = json!({"s": "a\"b\\c\nd\te\u{0001}f"});
        let c = String::from_utf8(canonical_json_bytes(&v).unwrap()).unwrap();
        assert_eq!(c, "{\"s\":\"a\\\"b\\\\c\\nd\\te\\u0001f\"}");
    }

    #[test]
    fn preserves_unicode_raw() {
        let v = json!({"n": "café ☕"});
        let c = String::from_utf8(canonical_json_bytes(&v).unwrap()).unwrap();
        assert_eq!(c, "{\"n\":\"café ☕\"}");
    }

    #[test]
    fn rejects_floats() {
        let v = json!({"x": 1.5});
        assert!(canonical_json_bytes(&v).is_err());
    }

    #[test]
    fn accepts_integer_extremes() {
        let v = json!({"a": i64::MIN, "b": u64::MAX});
        let c = String::from_utf8(canonical_json_bytes(&v).unwrap()).unwrap();
        assert_eq!(c, format!("{{\"a\":{},\"b\":{}}}", i64::MIN, u64::MAX));
    }

    #[test]
    fn deterministic_across_input_orderings() {
        let a: Value = serde_json::from_str(r#"{"x":1,"y":2}"#).unwrap();
        let b: Value = serde_json::from_str(r#"{"y":2,"x":1}"#).unwrap();
        assert_eq!(canonical_json_bytes(&a), canonical_json_bytes(&b));
    }
}

File: reference/python/murl_ref/canonical.py
"""MCF-1, the mURL canonical form — spec 7.1.

This is the byte form that hashes and signatures cover. Everything about it is
chosen to be boring:

* object members sorted by Unicode code point of the member name;
* no insignificant whitespace;
* the seven named escapes, remaining control characters as ``\\u00xx``, and
  everything else raw UTF-8;
* integers only, in plain decimal.

The float rejection is the load-bearing part. MCF-1 is byte-identical to
RFC 8785 (JCS) for every document the manifest schema allows, and it gets
there by removing the one piece of JCS that is genuinely hard to reimplement:
ECMAScript number formatting. A canonical form that is easy to get subtly
wrong produces signatures that break for reasons nobody can reproduce.

Two details the prose in 7.1 does not pin down, resolved here to match the
reference implementation (see the README's "Spec questions this raised"):

* the hex in ``\\u00xx`` is **lowercase**, as RFC 8785 requires;
* only code points below 0x20 are escaped -- U+007F is emitted raw, again as
  RFC 8785 does.
"""

from __future__ import annotations

from typing import Any, Dict

__all__ = [
    "CanonicalError",
    "canonicalize",
    "canonical_bytes",
    "signing_bytes",
    "INT_MIN",
    "INT_MAX",
]


class CanonicalError(ValueError):
    """The value cannot be represented in MCF-1."""


INT_MIN = -(2**63)
INT_MAX = 2**64 - 1

_SHORT_ESCAPES = {
    '"': '\\"',
    "\\": "\\\\",
    "\b": "\\b",
    "\f": "\\f",
    "\n": "\\n",
    "\r": "\\r",
    "\t": "\\t",
}


def _write_string(text: str, out: list) -> None:
    out.append('"')
    for char in text:
        escape = _SHORT_ESCAPES.get(char)
        if escape is not None:
            out.append(escape)
        elif ord(char) < 0x20:
            out.append("\\u%04x" % ord(char))
        else:
            out.append(char)
    out.append('"')


def _write(value: Any, out: list) -> None:
    if value is None:
        out.append("null")
    elif value is True:
        out.append("true")
    elif value is False:
        out.append("false")
    elif isinstance(value, int):
        # bool was handled above; Python's bool-is-int would otherwise emit 1/0.
        if not INT_MIN <= value <= INT_MAX:
            raise CanonicalError(f"integer {value} does not fit i64/u64")
        out.append(str(value))
    elif isinstance(value, float):
        raise CanonicalError(f"canonical form forbids non-integer numbers ({value})")
    elif isinstance(value, str):
        _write_string(value, out)
    elif isinstance(value, (list, tuple)):
        out.append("[")
        for index, item in enumerate(value):
            if index:
                out.append(",")
            _write(item, out)
        out.append("]")
    elif isinstance(value, dict):
        out.append("{")
        # Python compares str by code point, which is the ordering 7.1 asks
        # for. (RFC 8785 sorts by UTF-16 code unit; the two agree for every
        # member name below U+10000, which is every name the schema permits.)
        for index, key in enumerate(sorted(value)):
            if not isinstance(key, str):
                raise CanonicalError(f"object member name {key!r} is not a string")
            if index:
                out.append(",")
            _write_string(key, out)
            out.append(":")
            _write(value[key], out)
        out.append("}")
    else:
        raise CanonicalError(f"value of type {type(value).__name__} is not JSON")


def canonicalize(value: Any) -> str:
    """Return the MCF-1 text for a JSON value."""
    out: list = []
    _write(value, out)
    return "".join(out)


def canonical_bytes(value: Any) -> bytes:
    """Return the MCF-1 bytes for a JSON value. This is what gets hashed."""
    return canonicalize(value).encode("utf-8")


def signing_bytes(document: Dict[str, Any]) -> bytes:
    """The bytes a signature covers (7.2): the manifest minus ``signature``.

    Note what is *not* removed: unknown members. Signatures cover them (5.1),
    so a consumer that ignores a member still verifies the bytes that carried
    it. This function only prepares those bytes -- it does no cryptography.
    """
    if not isinstance(document, dict):
        raise CanonicalError("a manifest document must be an object")
    stripped = {k: v for k, v in document.items() if k != "signature"}
    return canonical_bytes(stripped)
